internal/services/chaosstudio/chaos_studio_experiment_resource.go

```diff
@@ -0,0 +1,588 @@
+	Steps             []StepSchema                               `tfschema:"steps"`
+	// tags are not fully supported yet, you can send them to the API, but they won't be returned
+	// Tags              map[string]interface{}                     `tfschema:"tags"`
+}
+
+type SelectorSchema struct {
+	Name      string   `tfschema:"name"`
+	TargetIds []string `tfschema:"chaos_studio_target_ids"`
+}
+
+type StepSchema struct {
+	Branch []BranchSchema `tfschema:"branch"`
+	Name   string         `tfschema:"name"`
+}
+
+type BranchSchema struct {
+	Actions []ActionSchema `tfschema:"actions"`
+	Name    string         `tfschema:"name"`
+}
+
+type ActionSchema struct {
+	ActionType   string            `tfschema:"action_type"`
+	SelectorName string            `tfschema:"selector_name"`
+	Duration     string            `tfschema:"duration"`
+	Urn          string            `tfschema:"urn"`
+	Parameters   map[string]string `tfschema:"parameters"`
+}
+
+func (r ChaosStudioExperimentResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return experiments.ValidateExperimentID
+}
+
+func (r ChaosStudioExperimentResource) ResourceType() string {
+	return "azurerm_chaos_studio_experiment"
+}
+
+func (r ChaosStudioExperimentResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"location": commonschema.Location(),
+		"name": {
+			ForceNew: true,
+			Required: true,
+			Type:     pluginsdk.TypeString,
+		},
+		"resource_group_name": commonschema.ResourceGroupName(),
+		"selectors": {
+			Required: true,
+			Type:     pluginsdk.TypeList,
+			MinItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"name": {
+						Required:     true,
+						Type:         pluginsdk.TypeString,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+					"chaos_studio_target_ids": {
+						Required: true,
+						Type:     pluginsdk.TypeList,
+						Elem: &pluginsdk.Schema{
+							Type:         pluginsdk.TypeString,
+							ValidateFunc: commonids.ValidateChaosStudioTargetID,
+						},
+					},
+				},
+			},
+		},
+		"steps": {
+			Required: true,
+			Type:     pluginsdk.TypeList,
+			MinItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"name": {
+						Required:     true,
+						Type:         pluginsdk.TypeString,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+					"branch": {
+						Required: true,
+						Type:     pluginsdk.TypeList,
+						MinItems: 1,
+						Elem: &pluginsdk.Resource{
+							Schema: map[string]*pluginsdk.Schema{
+								"name": {
+									Required:     true,
+									Type:         pluginsdk.TypeString,
+									ValidateFunc: validation.StringIsNotEmpty,
+								},
+								"actions": {
+									Required: true,
+									Type:     pluginsdk.TypeList,
+									MinItems: 1,
+									Elem: &pluginsdk.Resource{
+										Schema: map[string]*pluginsdk.Schema{
+											"action_type": {
```

Comment:
What if we made these into their own actions so we can validate at plan time rather than runtime?

Reply:
@mbfrahry that was my original approach, I had a `continuous`, `delay` and `discrete` block in the `actions` block, however multiple `continuous`, `delay` and `discrete` blocks can be specified in any order and this order has to be preserved. When I split them into their own types I lost the ordering as soon as multiple action types were supplied.